Unit 4.2 localStorage API

frankiebarrios opened the pull request:
## What It Does
This pull request is set to complete Unit 4.2 of the ux-onboarding-project.

This work is taking place to complete unit 4.2 of the ux-onboarding-project.

**Implementation:**
- Created required functions for localStorage API.

[closes #18 ]

## How To Test
- Install dependencies using instruction in README.
- Start server using yarn start or build using yarn build.

## Notes/Caveats
Looks like utilizing some packages (faker.js and linq.js) is causing warnings from webpack
upon compiling due to performance issues. These packages
are only in place for testing (getting me names for the new users)
and will be removed in next unit since I will be receiving input from the forms instead.

Looked into lazy loading/ code-splitting to resolve this
issue but didn't wanna go too far on this 1 unit. May 
be a good thing for me to try and utilize in the final unit as 
one of my 'mini projects'.
## Checklist

Under penalty of public shaming, I avow that I:
**Many items in list N/A** 
- [X] Reviewed the diff to look for typos, whitespace errors, and console.log() statements
- [X] Verified that there are no compiler or linter errors or warnings
- [X] Verified the build in production(optimized) mode
- [X] Updated the docs, if necessary
- [X] Included the appropriate labels
- [X] Referenced applicable issues (i.e. Closes #XXXX, Fixes #XXXX)

Browsers tested:
**Many items in list N/A** 
- [X] Chrome
- [x] Firefox
- [x] Safari
- [ ] Edge
- [ ] Internet Explorer

## Dependencies
1. Install Node and npm
2. Install Yarn
3. Install Ux-Lint

Commit:
Implemented one way data binding.

Commit:
Implemented one-way data binding.

Commit:
Working on adding localStorage.

Commit:
Clean up.

Commit:
Clean Up

Commit:
Converted lcoalStorage to Class and handled importing/ exporting to components.

Commit:
Checkpoint -- Implementing lifecycle methods and refining/ refactoring some LocalStorage methods.

Commit:
Checkpoint -- Majority of methods in the LocalStorage class are completed and tested successfully.

Commit:
Refactored createUser() and some other general clean up.

Commit:
Refactored some functions, added linq.js, and clean up.

Commit:
Unit 4.2 completed.

Commit:
Clean up

Commit:
Clean Up

kddrake commented on src/api/localStorage.js:
```diff
@@ -0,0 +1,111 @@
+import faker from '../../node_modules/faker';
+export class LocalStorage {
+  constructor(_id) {
+    this._id = _id;
```
Don't think having a class property for `_id` is needed, and then consequently the constructor itself. Have the functions work with passed in ids rather than `this._id`

kddrake replied:
Or, with that said, consider using the `constructor` to create/retrieve and parse the user information so that you can skip that step within your functions.

kddrake commented on src/components/create-user.html:
```diff
@@ -1,95 +1,157 @@
+      }
+      
+      createUser(user) {
+        localStorage.createUser(localStorage);
```
you'll want to keep all `localStorage` work from within your `LocalStorage` module. Or, more directly, you shouldn't be using `localStorage` in your components.

frankiebarrios replied:
Are we talking about the passing of localStorage in the calls or instantiating it and using the functions like that all together within the component?

frankiebarrios replied:
What I was doing to test my functions from LocalStorage was importing it to my component, making an instance of it (to be able to use it's methods), and then calling them like that. Is this the wrong way to go about importing and using the Class you're importing?

kddrake replied:
>  Is this the wrong way to go about importing and using the Class you're importing?

Nope, not at all. I just want your components to actually use your `LocalStorage` class to interface with `localStorage`, not `localStorage` directly. So, here in `create-user.html`, you'd call something like `LocalStorage.createUser(user)` once you've gotten all the user information data together into an object (or something). Your `LocalStorage` class then would do all the work to get it into the list of users and update `localStorage` so that the create user process is "saved".

kddrake replied:
> Are we talking about the passing of localStorage in the calls...

I wasn't initially but you certainly won't want to be doing that.

> ... or instantiating it and using the functions like that all together within the component?

Yeah, you'll still want to import it and create an instance of the class for your app to use.

frankiebarrios replied:
Made some changes to create-user.html and how the createUser function is being utilized and how to user object being passed into that function is being built. Can you take a look and see if these changes are valid according to what we were discussing previously ^?

frankielb replied:
Or should I set the properties like this and pass those props into the function calls?

![image](https://user-images.githubusercontent.com/51921679/61394666-3da94480-a889-11e9-8b0c-9fa14600fa80.png)

frankielb replied:
This is just an example from the docs I'm looking through. I'd modify it to my needs for the project of course.

collincahill replied:
I _think_ the `createUser` call is in much better shape now. You'll still need to remove any direct references to `localStorage` from the component (the only thing we want to access `localStorage` directly is the API).

frankiebarrios replied:
Ok great, I didn't do any refactoring to anything but the createUser(), wanted to make sure I did it right there before moving on to the rest of the app. Thanks again for the help!

kddrake commented on src/api/localStorage.js:
```diff
@@ -0,0 +1,111 @@
+    this._id = _id;
+  }
+
+  createUser() {
```
I would expect a user object (sans `_id`) to be passed into `createUser` to be used for the `userEntry`

frankiebarrios replied:
In localStorage.js I am creating the user to be pushed into localStorage within createUser(), should I create the user in the CreateUser Class using it's constructor and pass that in when I call createUser() from the CreateUser component?

kddrake replied:
> In localStorage.js I am creating the user to be pushed into localStorage within createUser()

No, you'll create the user in your components (probably the create user one). That's where the inputs for the user data are, so that's where you should build out the bulk of the user object. 

The `_id` property can be added there as well, but since that's not something the form itself is concerned with, it would also make sense for you to add the `_id` param in your LocalStorage class function `createUser()`, before stringifying and sending it to `localStorage`.

> should I create the user in the CreateUser Class using it's constructor...

Probably not. You'll want to use the same `LocalStorage` instance throughout your app, not every time you need to create a new user. Plus, [you're already creating an `_id` in `createUser`](https://github.com/frankielb/frankie-barrios-ux-aop/pull/22/files/1dc882215786c38a6c448f061e24f05c3659d669#diff-a09ed0d6e500df7a8a529a6e5048a611R11) as it is. Rather than assigning it to `this._id`, assign it to the `userEntry._id` a few lines down.

> ... and pass that in when I call createUser() from the CreateUser component?

Like I said in the first answer, you'll want to have the component create the bulk of the user object and pass it into `createUser`. You won't need to pass anything into `createUser` that was given back from the initial creation of the `LocalStorage` instance. The instance should just give you the ability to use the class functions, for the most part.

frankiebarrios replied:
Good deal, thanks for such a detailed response. Helps a lot. Cutting a lot of the 'fluff' out of a majority of my functions as well with this refactor.

kddrake commented on src/api/localStorage.js:
```diff
@@ -0,0 +1,111 @@
+
+  replaceUser(id, update) {
+    let users = JSON.parse(localStorage.getItem('userAccounts'));
+    for (let i = 0; i < users.length; i++) {
```
should use the simplified syntax
```suggestion
    users.forEach({
```

kddrake replied:
Better yet
```suggestion
    const index = users.indexOf(users.find(user => user._id === id));
    Object.assign(users[index], update);
```

frankiebarrios committed:
Update src/api/localStorage.js

Co-Authored-By: Kean Drake <kean.drake@banno.com>

Commit:
no message

Commit:
Merge branch 'Unit-4.2-localStorage-API' of github.com:frankielb/frankie-barrios-ux-iop into Unit-4.2-localStorage-API

# Conflicts:
#	src/api/localStorage.js

Commit:
Refactored some LocalStorage functions.

Commit:
Further refactoring of functions.

Commit:
Working on refactoring component to utilize LocalStorage from within component. Only createUser() has been refactored in this commit.

Commit:
Resolved issue with getUserIndex() not returning a value, resolved issue with createUser() not generating a new id with each new user, refactored function calls to utilize instance of API within component, refactored userEntry being used for function calls, and some general clean up.

kddrake commented on src/components/create-user.html:
```diff
@@ -1,95 +1,159 @@
+
+    class CreateUser extends PolymerElement {
+      static get is() { return 'create-user'; }
+      constructor() {
```
Since you're not actually doing anything with the constructor, let's nix it.

kddrake commented on src/components/create-user.html:
```diff
@@ -1,95 +1,159 @@
+      }
+      
+      createUser() {
+        LocalUsers.createUser(userEntry);
```
~Where is `userEntry` coming from?~ You should use the data from your inputs to populate the `userEntry` object. 

There's a good chance you won't get the information from the `on-click` event you need, so I'd suggest setting up a `userEntry` polymer property and passing it to `LocalUsers.createUser`. You'd also have to ensure that `userEntry` is being populated correctly, but that's a bit out of scope for `createUser` specifically.

kddrake commented on src/components/create-user.html:
```diff
@@ -1,95 +1,159 @@
+        <div class="form-border">
+          <form>
+            <label for="fname">First Name</label>
+            <input type="text" id="fname" name="firstname" placeholder="Your name..">
```
You'll want your inputs to update _something_, otherwise what're they doin'? Check out the **Two-way binding to a non-Polymer element** section of the [data binding docs](https://polymer-library.polymer-project.org/1.0/docs/devguide/data-binding). As I mentioned in another comment, you'll need to set up some sort of polymer property for this to work properly.

kddrake commented on src/components/create-user.html:
```diff
@@ -1,95 +1,159 @@
+        LocalUsers.createUser(userEntry);
+      }
+
+      getUser() {
```
This is the `create-user` component... why do you need a `getUser` function? Same goes for `getAllUsers`, `updateUser`, `replaceUser`, `deleteUser`, `deleteAllUsers`, `validateUsers`, and `testLocalStorage`. I understand you're testing your `LocalUsers` class but you should have these in their correct components if they're going to be merged.

frankiebarrios replied:
Yeah was basically using it for testing and making sure all functions ran correctly in one place for 4.2 and was going to start applying to correct components in 4.3.

Commit:
Removed unused Constructor.

Commit:
Removed  packages linq and faker, data binding for create-user.html, refactored new user being passed into createUser() using Polymer property, and cleaned up create-user.html.

kddrake commented on src/components/create-user.html:
```diff
@@ -1,95 +1,111 @@
+                lastName: '',
+                phone: '',
+                email: ''
+      } } } } }
```
nit: this is akward... though I don't know if that's just because we don't normally condense closing brackets all in one line

frankiebarrios replied:
Lol yeah it is weird, will change. It's a setting I accidentally left on in my formatters JSON file I keep.

kddrake commented on src/components/create-user.html:
```diff
@@ -1,95 +1,111 @@
+        <div class="form-border">
+          <form>
+            <label for="fname">First Name</label>
+            <input type="text" id="fname" name="firstname" placeholder="Your name.." value="{{user.firstName::input}}">
```
tiny nit: `id` and `name` aren't necessary. Fine to have, but you'll eventually notice we don't apply them unless we need to. And in this case, I don't believe you need to.

kddrake commented on src/api/localStorage.js:
```diff
@@ -0,0 +1,93 @@
+  }
+
+  createUser(user) {
+    if (this.users === null) this.users = [];
```
I would actually do this in your constructor... if you haven't already check out the comment there.

kddrake commented on src/api/localStorage.js:
```diff
@@ -0,0 +1,93 @@
+export class LocalStorage {
+  constructor() {
+    this.users = JSON.parse(localStorage.getItem('userAccounts'));
```
Good use of the `constructor`, I would go a bit further though. You can't be certain you'll have anything to parse, so you'll want to account for that.
```suggestion
    const storageUsers = localStorage.getItem('userAccounts');
    this.users = storageUsers ? JSON.parse(storageUsers) : [];
```
if you're not familiar with the **_prop_ ? _statement_ : _statement_** syntax, give [this](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Conditional_Operator) a look over. Conditional operator/ternary expressions are super handy. That said, you could also (probably) simplify it further by just using an `||` statement
```suggestion
    this.users = JSON.parse(localStorage.getItem('userAccounts')) || [];
```
This _should_ evaluate the first expression (`JSON.parse(localStorage.getItem('userAccounts'))`), and if false, sets `this.users` to the expression on the other side of the `||`. Also pretty nifty.

kddrake commented on src/api/localStorage.js:
```diff
@@ -0,0 +1,93 @@
+    if (this.users === null) this.users = [];
+    user._id = Math.floor(10000 + Math.random() * 9000);
+    this.users.push(user);
+    localStorage.setItem('UserAccounts',
```
you'll have to make sure your `getItem` and `setItem` naming is consistent. I would think with the way you've got it set up now, you're always setting `UserAccounts` but getting from `userAccounts`

kddrake replied:
Maybe create a private `updateUsers` function to call when you need to update `localStorage` with the freshest `this.users`. I would assume this'll need to happen in a few of these class functions.

kddrake reviewed (changes requested):
Create user is looking a lot better component side, got a few comments on the LocalStoarge class though

Commit:
Refactored localStorage constructor and createUser function, cleaned up form attributes not being used, formatted code.

Commit:
Clean Up

Commit:
Checkpoint: user-list working with hardcoded values only at this time. Clean up.

Commit:
Having trouble utilizing user list coming in from getAllUsers() within my component.

kddrake commented on src/components/user-list.html:
```diff
@@ -30,18 +31,21 @@ <h2> User list: </h2>
       static get properties() {
         return {
           users: {
             value() {
```
I think you'll want
```
value: () => LocalUsers.getAllUsers()
```

Commit:
no message

kddrake commented on src/api/localStorage.js:
```diff
@@ -0,0 +1,90 @@
+  }
+
+  createUser(user) {
+    this.users.push(JSON.stringify(user));
```
as discussed in slack, you'll want to just push the user object **without** stringifying it.
```suggestion
    this.users.push(user);
```

you'll still want to stringify `this.users` when setting `UserAccounts`, though. just don't need the double `stringify`.

kddrake commented on src/components/intern-app.html:
```diff
@@ -31,24 +31,44 @@
+    const LocalUsers = new LocalStorage();
+
     class InternApp extends PolymerElement {
+      constructor() {
```
nit: move this below any of your `static get` functions, but before any other component functions.

kddrake commented on src/components/intern-app.html:
```diff
@@ -31,24 +31,44 @@
+        return {
+          userAccounts: {
+            type: Array,
+            value: () => {
```
You can forego the constructor work and just use
```suggestion
            value: () => LocalUsers.getAllUsers()
```

kddrake commented on src/components/intern-app.html:
```diff
@@ -31,24 +31,44 @@
-    
+    import { LocalStorage } from '../api/localStorage.js';
+
+    const LocalUsers = new LocalStorage();
```
since this is technically a [Storage](https://developer.mozilla.org/en-US/docs/Web/API/Storage) object, you may want to rename this to reflect that.

kddrake commented on src/components/user-list-test.html:
```diff
@@ -0,0 +1,49 @@
+      <template>
+        <div>
+          <user-profile 
+            id="{{item.id}}" 
```
All of these properties should be one-way bound (`[[ ... ]]`) instead of two-way bound (`{{ ... }}`)
```suggestion
            id="[[item.id]]" 
```

About the only time you'll _ever_ use two way binding is when you have a text input that you want to update some model. That's why you're two-way binding in your create and edit user components. Here though (and just about everywhere else)? You'll want to one-way bind.

kddrake commented on src/components/user-list-test.html:
```diff
@@ -0,0 +1,49 @@
+  </style>
+  <template>
+    <h2> User list: </h2>
+    <dom-repeat items="{{users}}">
```
I'm sure this works, however, you should get used to doing it this way:
```
<template is="dom-repeat" items="[[users]]>
  <div>
    <user-profile ...></user-profile>
  </div>
</template>
```
Beyond "it's what we do", there's a good reason for using the `template is="..."` syntax over `dom-repeat`. At the moment it escapes me, though.

kddrake replied:
Also needs to be one-way bound, like the properties on `user-profile`

kddrake commented on src/components/user-list-test.html:
```diff
@@ -0,0 +1,49 @@
+        return {
+          users: {
+            value() {
+              return [
```
Don't need this since you're populating and passing users into the component. Just need to declare it's type.
```suggestion
             users: Array,
```

kddrake commented on src/components/user-list.html:
```diff
@@ -1,24 +1,54 @@
+      <template>
+        <div>
+          <user-profile 
+            id="{{item.id}}" 
```
One-way bind these, too

kddrake replied:
Ah, just make all the changes I suggested in `user-list-test` here, and then get rid of the test component once this is good to go.

Commit:
Made all changes recommended in latest PR conversations:
- Made all data binding one way, instead of inputs.
- Removed user-list-test
- Removed constructor, properties from intern-app (not being used)
- Refactored some Polymer Properties
- Clean Up

Commit:
Refactored dom-repeat to Polymer 1 version.

Commit:
Resolved Conflicts

Commit:
Merge branch 'master' of github.com:frankielb/frankie-barrios-ux-iop into Unit-4.2-localStorage-API

# Conflicts:
#	src/components/intern-app.html

Commit:
Working on getting edit user working correctly. Having issue with editUser() placing each individual char into localStorage object.

Commit:
no message

Commit:
Resolved issue with editUser().

Commit:
Resolved issue with deleteUser() in localStorage.

Commit:
Resolved issue with createUser(). Made changes to how id is being generated and broke edit and delete user functions.

Commit:
Resolved issues with deleteUser() and editUser()

Commit:
Clean Up and page reload change.

Commit:
Removed vscode settings and added vscode to .gitignore.

Commit:
Removed unused dir.

Jropp commented on src/api/localStorage.js:
```diff
@@ -0,0 +1,110 @@
+    this.updateUsers();
+    user.id = this.generateId();
+    this.users.push(user);
+    localStorage.setItem('UserAccounts', JSON.stringify(this.users));
```
Since you are already setting and updating a this.users object, could you just add an updateDatabase function that sets the database to be the current this.users property?

SoftDevFb replied:
Not quite sure what we're talking about here. So for this function ^ I am pulling in the current  this.users with the this.updateUsers() to make sure I have the most up to date version of this.users since I have it setting in my constructor, then I am setting my id, pushing the new user with the new id to the current version of this.users and then setting that to the database. Do you mean like a helper that does the localStorage setting for me?

Jropp replied:
Exactly. 

> Do you mean like a helper that does the localStorage setting for me?

frankiebarrios replied:
Complete

Jropp commented on src/api/localStorage.js:
```diff
@@ -0,0 +1,110 @@
+    this.users.push(user);
+    localStorage.setItem('UserAccounts', JSON.stringify(this.users));
+    console.log('User Added: ', user);
+    return this.users;
```
This turns createUser into a createUserAndReturnUsers function. I'd just leave this return out and any time you need to get the users in your view component just make a call to getAllUsers.

SoftDevFb replied:
Complete

Jropp commented on src/api/localStorage.js:
```diff
@@ -0,0 +1,110 @@
+
+  updateUsers() {
+    this.users = JSON.parse(localStorage.getItem('UserAccounts')) || [];
+    return this.users;
```
same as the comment above

SoftDevFb replied:
Complete

Jropp commented on src/api/localStorage.js:
```diff
@@ -0,0 +1,110 @@
+    try {
+      const index = this.getUserIndex(id);
+      const user = this.users[index];
+      return user;
```
You could condense this process down to a single line with array.find since you are basically finding and returning an object, but consider that a suggestion rather than a requested change. 

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find

SoftDevFb replied:
Great tip, worked awesomely! Gonna use that one from now on for similar processes.
    const user = this.users.find(user => String(user.id) === String(id));
Along with the String() conversion. Have always used toString() as my go to but like that one more.

Jropp commented on src/api/localStorage.js:
```diff
@@ -0,0 +1,110 @@
+  }
+ 
+  getUser(id) {
+    const user = this.findUserMatch(id);
```
Actually, (see comment below) I'd just use array.find on this.users and delete findUserMatch altogether

Jropp commented on src/api/localStorage.js:
```diff
@@ -0,0 +1,110 @@
+
+  replaceUser(id, update) {
+    const index = this.getUserIndex(id);
+    const replacedUser = Object.assign(
```
Could you just put this on a single line?
```
   const replacedUser = Object.assign(this.users[index], update);
```

SoftDevFb replied:
Used that same Array.find here to get my user and just passed that in here too:
    const user = this.users.find(user => String(user.id) === String(id));
    const replacedUser = Object.assign(user, update);

Jropp commented on src/api/localStorage.js:
```diff
@@ -0,0 +1,110 @@
+  deleteUser(id) {
+    this.updateUsers();
+    this.users.forEach((user, index) => {
+      if (user.id.toString() === id.toString()) {
```
Super nitpick but you _could_ also use 

```
String(user.id) === String(id)
```

collincahill replied:
Are the `id` properties not already stored as strings?

SoftDevFb replied:
Not at the moment. I have the generateId() running from createUser() to generate a random id each time a user is created. Was creating it in the component props initially but was having an issue with the same id being used for every user.

Commit:
Refactoring within deleteUser(), removed returns from createUser(), and updateUser().

Commit:
Refactored per PR convo.

Jropp commented on src/api/localStorage.js:
```diff
@@ -0,0 +1,97 @@
+    this.users.forEach((user, index) => {
+      if (String(user.id) === String(id)) {
+        this.users.splice(index, 1);
+        localStorage.setItem('UserAccounts', JSON.stringify(this.users));
```
You could use your new helper function here

frankiebarrios replied:
Completed

Commit:
Created and implemented helper function to set local data.

Commit:
Added UI change to <user-list>/ <user-profile> to display username and id when the user card is minimized.

Commit:
Fixed issue with editUser() not replacing user with update user, but duplicating user instead. Also added some error catching and general clean up.

kddrake commented on src/api/localStorage.js:
```diff
@@ -0,0 +1,96 @@
+  }
+
+  createUser(user) {
+    this.updateUsers();
```
Rather than updating the user list before adding/deleting a user (and having it in it's own function), run it after `localStorage.setItem('UserAccounts', JSON.stringify(this.users));` in `setLocalStorageData()`. That way, `this.users` will _always_ be up to date with your local storage, and you won't have to call it but once.

... maybe rename the function too, then. something like `syncStorage`, so that it's clear you're doing more than just setting localStorage.

kddrake commented on src/api/localStorage.js:
```diff
@@ -0,0 +1,96 @@
+      const updatedUser = Object.assign(this.getUser(id), editObj);
+      this.deleteUser(id);
+      this.createUser(updatedUser);
+    } catch (err) {console.log('Error Editing User: ', err);}
```
I'd probably go with this, instead:
```suggestion
    } catch (err) {console.error('Error Editing User: ', err);}
```

Seems to be a decent reason to throw an error.

kddrake commented on src/api/localStorage.js:
```diff
@@ -0,0 +1,96 @@
+    try{
+      const editObj = {};
+      editObj[`${editName}`] = `${editValue}`;
+      const updatedUser = Object.assign(this.getUser(id), editObj);
```
could be a bit simpler syntactically (maybe)
```suggestion
      const updatedUser = Object.assign(this.getUser(id), {editName: editValue});
```
and then you'd be able to remove the two lines above creating `editObj`

frankiebarrios replied:
Got this to work:      
const updatedUser = Object.assign(this.getUser(id), {[editName]: editValue});

kddrake commented on src/api/localStorage.js:
```diff
@@ -0,0 +1,96 @@
+      const editObj = {};
+      editObj[`${editName}`] = `${editValue}`;
+      const updatedUser = Object.assign(this.getUser(id), editObj);
+      this.deleteUser(id);
```
Although this is pretty clever
```
      this.deleteUser(id);
      this.createUser(updatedUser);
```
you could not only replace the user with a single line, but you wouldn't have to set `localStoarge` twice for a single action

```suggestion
      this.users.splice(this.users.findIndex(user => user.id === id), 1, updatedUser);
      this.syncStorage();
```

frankiebarrios replied:
Not sure if this one will work since I am using createUser() to generate my user id. Are you saying to remove the calls to deleteUser() and createUser() or just deleteUser()? Confused cause I see we're passing in the updated user to the splice.

kddrake replied:
why do you need to generate an id if you're only editing a user?

kddrake replied:
> Are you saying to remove the calls to deleteUser() and createUser() or just deleteUser()?

remove both

frankiebarrios replied:
I see so the updatedUser will keep the current users id. Got it, thank you.

kddrake commented on src/api/localStorage.js:
```diff
@@ -0,0 +1,96 @@
+    } catch (err) {console.log('Error Editing User: ', err);}
+  }
+
+  replaceUser(id, update) {
```
what's this being used for? And if nothing, can it be canned?

kddrake commented on src/api/localStorage.js:
```diff
@@ -0,0 +1,96 @@
+
+  replaceUser(id, update) {
+    const user = this.users.find(user => String(user.id) === String(id));
+    const replacedUser = Object.assign(user, update);
```
if you're not doing anything with it beyond returning, skip the `const` assignment and just return it
```suggestion
    return Object.assign(user, update);
```

kddrake commented on src/api/localStorage.js:
```diff
@@ -0,0 +1,96 @@
+
+  deleteUser(id) {
+    this.updateUsers();
+    this.users.forEach((user, index) => {
```
The loop is odd to me, though I don't know I'd say it's any less efficient than what I'd do. I'll give it to you anyway and you can see what you think.

```suggestion
    const userIndex = this.users.indexOf(this.users.find(user => user.id === id));
    this.users.splice(userIndex, 1);
    this.setLocalStorageData();
```

kddrake commented on src/api/localStorage.js:
```diff
@@ -0,0 +1,96 @@
+  }
+
+  replaceUser(id, update) {
+    const user = this.users.find(user => String(user.id) === String(id));
```
Are the `String(...)` wrappings necessary? If not i'd ditch'm

kddrake commented on src/api/localStorage.js:
```diff
@@ -0,0 +1,96 @@
+      _id: String,
+      name: String
+    }
+    const userObject = Object.getOwnPropertyNames(user);
```
`userObject` is an array of property names, correct? It's a little deceiving that the prop is called `userObject`, then, because it's not actually an Object. Something like `userProps` would be more fitting.

Of course, same goes for `testObject`

kddrake commented on src/api/localStorage.js:
```diff
@@ -0,0 +1,96 @@
+    }
+    const userObject = Object.getOwnPropertyNames(user);
+    const testObject = Object.getOwnPropertyNames(validationObject);
+    if (userObject.length != testObject.length) {
```
```suggestion
    if (userObject.length !== testObject.length) {
```

kddrake commented on src/api/localStorage.js:
```diff
@@ -0,0 +1,96 @@
+      return false;
+    }
+
+    for (let i = 0; i < userObject.length; i++) {
```
It's a little weird but this could be simplified
```suggestion
    return !Boolean(userObject.find(prop => userObject[prop] !== testObject[prop]));
```
- `.find()` loops through `userObject`. 
- if it finds one of the iterations passes `userObject[prop] !== testObject[prop]`, it'll return that value. if none of the iterations passes, it returns `undefined`
- `Boolean()` will convert _`someValueFromFind`_ into `true` and `undefined` to `false`, the opposite of what you need. So inverse it with `!` and return

kddrake commented on src/api/localStorage.js:
```diff
@@ -0,0 +1,96 @@
+
+  getUserIndex(id) {
+    try {
+      const index = this.users.indexOf(this.users.find(user => user.id.toString() === id.toString()));
```
```suggestion
      const index = this.users.findIndex(user => user.id === id);
```

kddrake commented on src/api/localStorage.js:
```diff
@@ -0,0 +1,96 @@
+      const index = this.users.indexOf(this.users.find(user => user.id.toString() === id.toString()));
+      return index;
+    } catch (err) {
+      console.log('Error Finding Index: ', err);
```
```suggestion
      console.error('Error Finding Index: ', err);
```

kddrake commented on src/components/create-user.html:
```diff
@@ -1,95 +1,125 @@
         }
+      }
+
+      getUserIndex() {
```
What is this being used for? If nothing, remove it please.

kddrake commented on src/components/create-user.html:
```diff
@@ -1,95 +1,125 @@
-        customElements.define(CreateUser.is, CreateUser);
+      createUser() {
+        Storage.createUser(this.user);
+        window.location.reload();
```
![... but why](http://giphygifs.s3.amazonaws.com/media/1M9fmo1WAFVK0/giphy.gif)

SoftDevFb replied:
Lol. Added to have to page reload after the new user was added to automatically generate the new user-list showing that added user. I'll take it out.

kddrake commented on src/components/edit-user.html:
```diff
@@ -1,95 +1,116 @@
+        <div class="form-border">
+          <form>
+            <label for="fname">Enter User Id</label>
+            <input type="text" placeholder="User Id for edit and delete functions tests.." value="{{user.userId::input}}">
```
This is the _one_ thing you shouldn't allow a user to edit. ids should be constant from creation to deletion.

SoftDevFb replied:
Have this for the user to be able to choose which user they want to edit or delete.

kddrake replied:
Ok got this wrong. Disregard my original comment. However...
I think in the end, when you polish your project, you're not going to be super happy with this implementation. Maybe not, but that's my feeling.

For instance: **Remembering an id, navigating to edit user, inputing remembered id** has a _much_ greater chance for error than something like **Click edit button on user profile, user info pre-populates edit user component**. The latter method has virtually _no_ error potential (with the exception of a mis-click).

SoftDevFb replied:
I agree with ya, that would be a better way to go about it. I'll go ahead and get that going.

SoftDevFb replied:
Something like this?

![editUserRedesign](https://user-images.githubusercontent.com/29979289/62232458-e2e51200-b38b-11e9-9a35-53ed7ee7ab12.jpg)

SoftDevFb replied:
Still thinking of how to avoid issues with the 'edit field' part. For instance, if they want to edit 'firstName' and don't case the prop name right it'll add a second 'firstname' prop.

kddrake replied:
> Something like this?

Yessir. The only comment I would have is I would have the expand/collapse button down with the Edit User button. But that's just what I'd expect

frankiebarrios replied:
Sounds good. Thanks!

kddrake replied:
> For instance, if they want to edit 'firstName' and don't case the prop name right it'll add a second 'firstname' prop.

yeah that's just going to be an issue with any kind of "create a property" type logic, hence the static inputs that _only_ change a single prop.

you could modify the property name, once inputed, to camel case. check if the property is already created and notify the user the property already exist. I don't know how much of a rabbit hole that would end up being, but a potential solution none the less.

kddrake commented on src/components/edit-user.html:
```diff
@@ -1,95 +1,116 @@
+      }
 
-        customElements.define(EditUser.is, EditUser);
+      editUser() {
```
as we talked about in standup on Friday (7/26/19), you'll want to set up static inputs for the four main user properties, rather than actually allowing the user to create new fields. Or, alongside allowing users to create new fields if you want to keep the code around.

kddrake commented on src/components/user-profile.html:
```diff
@@ -1,111 +1,143 @@
+    class UserProfile extends PolymerElement {
+      static get is() { return 'user-profile'; }
+
+      expandCollapse() {
```
gross. this can be done in a way simpler fashion.
- set a property, something like `profileExpanded`, on your component
- wrap `phone` and `email` fields in `dom-if`s using the new property
- on `expandCollapse`, simply update the new property to the opposite of what it currently is

This new prop would also allow you to, pretty easily, update the wording of the button to display what action will actually happen, rather than `Expand/collapse`

kddrake commented on src/components/user-list.html:
```diff
@@ -1,24 +1,75 @@
+  <template>
+    <div class="centered">
+      <h2 class="centered">User List:</h2>
+      <button class="button" value="Refresh Page" onClick="window.location.reload();">Generate User List</button>
```
![... but why?](http://giphygifs.s3.amazonaws.com/media/1M9fmo1WAFVK0/giphy.gif)

shouldn't need to have the window reload just to get updated data.

kddrake commented on src/components/user-list.html:
```diff
@@ -1,24 +1,75 @@
+      </div>
     </template>
+  </template>
+  <button on-click="getUsers">Test getAllUsers()</button>
```
not sure what the needs for this is...

kddrake commented on src/components/user-list.html:
```diff
@@ -1,24 +1,75 @@
-    </script>
+      getUsers() {
+        Storage.getAllUsers();
+        console.log('this.users', this.users);
```
this is cool to have around locally, but we're not going to want to commit it to `master`.

Commit:
Changes Made:
- Refactored setLocalStorageDate() to include updateUsers() call and renamed to syncStorage.
- Logging errors more efficiently now throughout localStorage.
- Renamed userObject and testObject in validateUser() to userProps and testProps.

Commit:
- Refactored deleteUser().
- Changed String conversion method in getUserIndex().
- Refactored Object.assign usage in editUser().

Commit:
Refactored validateUser().

Commit:
- Removed getUserIndex().
- Refactored deleteUser().

Commit:
Refactored editUser().

Commit:
Clean Up

Commit:
Refactored editUser() call and function.

Commit:
- Renamed editUser() to updateUser().
- Clean Up

Commit:
- Resolved issue with random number generator in generateId().
- Resolved issue with import.

Commit:
- Refactored expandCollapse() to utilize dom-if.
- Refactored expand/ collapse button naming.
- Form clean up.

Commit:
- Refactored expandCollapse()
- Clean Up API

Commit:
- Styling changes.
- Changed button placement to trigger deleteAllUsers().

Commit:
- Styling changes.

Commit:
- Added round buttons.

Commit:
no message

Commit:
- Further styling work.

Commit:
- Styling changes and clean up.
- Refactored data binding in <user-list> and <user-profile>.

Commit:
- CSS Improvements

Commit:
- 'Delete' button on User Cards working

Commit:
Clean Up

Commit:
- Props and functions added to <user-profile> for new edit user functionality.

Commit:
- 'Cancel' Button on edit form working.
- Rough styles for edit form created.

Commit:
- Refactored updateUser() to work with new edit user functionality.
- Refactored <edit-user>

Commit:
- Clean Up.
- Added window reload upon running deleteUser() so new user list is automatically generated showing user deleted from the list.

kddrake commented on src/api/localStorage.js:
```diff
@@ -0,0 +1,68 @@
+
+  syncStorage() {
+    localStorage.setItem('UserAccounts', JSON.stringify(this.users));
+    this.updateUsers();
```
rather than having `updateUsers` in it's own function, just bring the logic into `syncStorage`
```suggestion
    this.users = JSON.parse(localStorage.getItem('UserAccounts')) || [];
```

if you were using `updateUsers` elsewhere in the file I would understand why it's in it's own function, however it's not, so it seems like it'd be better to consolidate.

kddrake commented on src/api/localStorage.js:
```diff
@@ -0,0 +1,68 @@
+  }
+
+  getUser(id) {
+    const user = this.users.find(user => String(user.id) === String(id));
```
since you're only assigning `user` to return it, just return without assigning to a `const`
```suggestion
    return this.users.find(user => String(user.id) === String(id));
```

kddrake commented on src/api/localStorage.js:
```diff
@@ -0,0 +1,68 @@
+  updateUser(id, update) {
+    try {
+      const updatedUser = Object.assign(this.getUser(id), update);
+      this.users.splice(
```
tiny nit: maybe pull `this.users.findIndex()` into a `const`, I think that'd make your `splice` function slightly more readable.

kddrake replied:
Or, since you need to grab the user index in a few spots, maybe pull it out into it's own function to be called by `updateUser` and `deleteUser`

kddrake commented on src/api/localStorage.js:
```diff
@@ -0,0 +1,68 @@
+  }
+
+  deleteUser(id) {
+    const userIndex = this.users.findIndex(user => String(user.id) === id);
```
see my comment about pulling this logic out into it's own function

kddrake commented on src/api/localStorage.js:
```diff
@@ -0,0 +1,68 @@
+
+  deleteAllUsers() {
+    this.users = [];
+    localStorage.clear();
```
tiny nit: you may want to ensure `localStorage.clear()` has run before resetting `this.users`, in the event the clear doesn't actually complete.

kddrake commented on src/components/create-user.html:
```diff
@@ -1,95 +1,136 @@
-    <script type="module">
+  <script type="module">
+    import { Element as PolymerElement } from '@banno/polymer/polymer-element.js';
+    import { LocalStorage } from '../api/localStorage.js';
```
this should be passed from `intern-app`, or whatever your most parent component is. There really shouldn't be a need to import  your LocalStorage api anywhere but `intern-app`

kddrake commented on src/components/create-user.html:
```diff
@@ -1,95 +1,136 @@
+      }
 
-        customElements.define(CreateUser.is, CreateUser);
+      getUserIndex() {
```
again, it doesn't look like you're using this function at all here. let's get rid of it.

kddrake commented on src/components/edit-user.html:
```diff
@@ -1,95 +1,142 @@
-    </template>
+        <button class="button" on-click="updateUser">Edit User</button>
+        <button class="button" on-click="deleteUser">Delete User</button>
+        <button class="button" on-click="deleteAllUsers">Delete All Users</button>
```
I don't think this "Delete all user" button belongs here. Makes more sense in the users list or intern app components.

kddrake commented on src/components/edit-user.html:
```diff
@@ -1,95 +1,142 @@
+            <input type="text" placeholder="Enter the user's id to delete or edit..." value="{{user.userId::input}}">
+
+            <label for="fname">Edit Name</label>
+            <input type="text" placeholder="Enter edit name..(key)" value="{{user.editName::input}}">
```
If you're going to keep going this way (and not updating the edit user to _also_ have static fields for the four properties), you're going to want to make sure you're properly formatting keys. Afaict, if you put `first name` here, your user object would have both `"first name"` and `firstName` properties.

frankiebarrios replied:
The code left within edit-user was from the initial attempt I made at edit-user where I misunderstood the docs on it. I left it there to possibly use that as an extra "feature" in the future but I'll clean it out of there. Have the edit-user functionality within the user-profile after refactoring that initial edit-user attempt. Now it's all handled on the user-profile card.

kddrake commented on src/components/intern-app.html:
```diff
@@ -33,16 +33,19 @@
     import { Element as PolymerElement } from '@banno/polymer/polymer-element.js';
+    import { LocalStorage } from '../api/localStorage.js';
+
+    const Storage = new LocalStorage();
```
this should be set as a polymer prop and bound to all the components that need to use the `LocalStorage`

kddrake commented on src/components/user-list.html:
```diff
@@ -1,24 +1,62 @@
+        };
+      }
 
+      getUsers() {
```
doesn't look like this is being used. let's nix it.

kddrake commented on src/components/user-profile.html:
```diff
@@ -1,111 +1,256 @@
+    </template>
+
+    <template is="dom-if" if="[[editButtonClicked]]">
+        <div class="edit-div edit-container edit-card">
```
nit: this section between the `<template is="dom-if" if="[[editButtonClicked]]">` could be moved into your edit user component, the component imported into `user-profile`, and used here rather hardcoding it within the user profile component itself.

kddrake commented on src/components/user-profile.html:
```diff
@@ -1,111 +1,256 @@
+      }
+
+      editButtonClick() {
+        this.editButtonClicked ? this.editButtonClicked = false :
```
```suggestion
        this.editButtonClicked = !this.editButtonClicked;
```

kddrake commented on src/components/user-profile.html:
```diff
@@ -1,111 +1,256 @@
+      }
+
+      deleteUser() {
+        Storage.deleteUser(`${this.user.id}`);
```
```suggestion
        Storage.deleteUser(this.user.id);
```

Commit:
Clean Up

Commit:
- Changed getUser() in localStorage API to return result directly.
- Small redesign of the 'Edit User'  section of <user-profile>.
- Removed unused getUsers() from user-list.
- Clean Up

Commit:
- Fixed an issue with validateUser()
- Refactored editButtonClick()
- Cleaned up some unused code from <user-profile>

Commit:
- Styling changes
- Clean up

Commit:
- Refactored <edit-user> to work from within the component being rendered form within <user-profile> vs hardcoding the work onto <user-profile> directly.
- Cancel button still needs work but other tests were successful.

Commit:
- Built helper to retrieve userIndex
- Global styling changes
- Refactored functions to utilize new getUserIndex( ) helper
- Refactored deleteAllUsers( ) to clear LS prior to resetting this.users
- Refactored cancel and edit button click functions

Commit:
Clean Up

Commit:
Refactored all data binding and removed all instantiating from components.
Now pulling in LocalStorage in Parent component and sharing all the way down through children.

kddrake commented on src/components/edit-user.html:
```diff
@@ -1,95 +1,146 @@
         }
+      }
+
+      editButtonClick() {
```
don't think you need this function

kddrake commented on src/components/user-list.html:
```diff
@@ -1,24 +1,60 @@
 
+      deleteAllUsers() {
+        this.storage.deleteAllUsers();
+        window.location.reload();
```
what's the purpose of calling `reload()`? I really don't think you should need to use it... about ever.

frankiebarrios replied:
Just have this here so a new user list is automatically generated to show the changes made vs having to reload the page. The data is being updated in localStorage in real time it just requires a reload to update the user cards in the list to reflect the change made in localStorage.

kddrake reviewed (changes requested):
just a few more things to address. this is looking really good 👍

Commit:
Clean Up

Commit:
- Refactored <create-user>, <user-list>'s 'Delete All Users' button, and <user-profiles>'s 'Delete User' button to utilize event bubbling to fix reload issue.
- Still working on <edit-user> to do the same.

Commit:
-Refactored <edit-user> to resolve reload issue. Now utilizing event bubbling for reloading app.

kddrake commented on src/components/intern-app.html:
```diff
@@ -31,18 +31,45 @@
+      }
+      
+      updateUserList(e) {
+        this.users = e.detail.data.updatedUsers;
```
rather than running `this.storage.getAllUsers()` before/while dispatching `updateUserList`, and then setting `this.users` to `e.detail.data.updatedUsers`, just watch for the call _here_ and then make the `this.storage.getAllUsers()` here as well. So everything in `intern-app` will stay the same, other than this line.

```suggestion
        this.users = this.storage.getAllUsers();
```

then your message just needs to dispatch, let the component catching it worry about updating the users list

```
        this.dispatchEvent(new CustomEvent('updateUserList', { bubbles: true, composed: true }));
```

frankiebarrios replied:
Done

Commit:
- Refactored events in all components to require components to just listen and then do the work within the component once the event is fired.
- Clean Up.

kddrake reviewed (approved):
Gonna add a UI cleanup in Jira here in a bit for ya, a few things I think could make for a better experience in the browser. That said:

This is looking real good. Nice job refining this 👍
